src/github.com/matrix-org/dendrite/roomserver/input/events.go

```diff
@@ -44,3 +61,224 @@ func processRoomEvent(db RoomEventDatabase, input api.InputRoomEvent) error {
+
+	// Load the events we need.
+	keysNeeded := stateKeysNeeded(result.stateNIDMap, needed)
+	var eventNIDs []int64
+	result.state = newStateEntryMap(state)
+	for _, keyNeeded := range keysNeeded {
+		eventNID, ok := result.state.lookup(keyNeeded)
+		if ok {
+			eventNIDs = append(eventNIDs, eventNID)
+		}
+	}
+	events, err := db.Events(eventNIDs)
+	if err != nil {
+		return
+	}
+	result.events = newEventMap(events)
+	return
+}
+
+// stateKeysNeeded works out which numeric state keys we need to authenticate some events.
+func stateKeysNeeded(stateNIDMap idMap, stateNeeded gomatrixserverlib.StateNeeded) []types.StateKeyTuple {
+	var keys []types.StateKeyTuple
+	if stateNeeded.Create {
+		keys = append(keys, types.StateKeyTuple{types.MRoomCreateNID, types.EmptyStateKeyNID})
+	}
+	if stateNeeded.PowerLevels {
+		keys = append(keys, types.StateKeyTuple{types.MRoomPowerLevelsNID, types.EmptyStateKeyNID})
+	}
+	if stateNeeded.JoinRules {
+		keys = append(keys, types.StateKeyTuple{types.MRoomJoinRulesNID, types.EmptyStateKeyNID})
+	}
+	for _, member := range stateNeeded.Member {
+		stateKeyNID, ok := stateNIDMap.lookup(member)
+		if ok {
+			keys = append(keys, types.StateKeyTuple{types.MRoomMemberNID, stateKeyNID})
+		}
+	}
+	for _, token := range stateNeeded.ThirdPartyInvite {
+		stateKeyNID, ok := stateNIDMap.lookup(token)
+		if ok {
+			keys = append(keys, types.StateKeyTuple{types.MRoomThirdPartyInviteNID, stateKeyNID})
+		}
+	}
+	return keys
+}
+
+type idMap map[string]int64
+
+func newIDMap(ids []types.IDPair) idMap {
+	result := make(map[string]int64)
+	for _, pair := range ids {
+		result[pair.ID] = pair.NID
+	}
+	return idMap(result)
+}
+
+// lookup an entry in the id map.
+func (m idMap) lookup(id string) (nid int64, ok bool) {
+	// Use a hash map here.
+	// We could use binary search here like we do for the maps below as it
+	// would be faster for small lists.
+	// However the benefits of binary search aren't as strong here and it's
+	// possible that we could encounter sets of pathological strings since
+	// the state keys are ultimately controlled by user input.
+	nid, ok = map[string]int64(m)[id]
+	return
+}
+
+type stateEntryMap []types.StateEntry
+
+// newStateEntryMap creates a map from a sorted list of state entries.
+func newStateEntryMap(stateEntries []types.StateEntry) stateEntryMap {
+	return stateEntryMap(stateEntries)
+}
+
+// lookup an entry in the event map.
+func (m stateEntryMap) lookup(stateKey types.StateKeyTuple) (eventNID int64, ok bool) {
+	// Since the list is sorted we can implement this using binary search.
+	// This is faster than using a hash map.
+	// We don't have to worry about pathological cases because the keys are fixed
+	// size and are controlled by us.
+	list := []types.StateEntry(m)
+	i := sort.Search(len(list), func(i int) bool {
+		return !list[i].StateKeyTuple.LessThan(stateKey)
+	})
+	if i < len(list) && list[i].StateKeyTuple == stateKey {
+		ok = true
+		eventNID = list[i].EventNID
+	}
+	return
```

Comment:
I have no confidence that this function looks up things correctly. Please add unit tests (should be trivial).